lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -980,41 +1058,317 @@ void crypto_cipher_final(void *ctx __unused, uint32_t algo __unused)
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result crypto_mac_alloc_ctx(void **ctx_ret, uint32_t algo)
+{
+	const mbedtls_md_info_t *md_info __maybe_unused;
+	int lmd_res __maybe_unused;
+	TEE_Result res = TEE_SUCCESS;
+	size_t ctx_size;
+	void *ctx;
+
+	res = mac_get_ctx_size(algo, &ctx_size);
+	if (res)
+		return res;
+
+	ctx = calloc(1, ctx_size);
+	if (!ctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		md_info = get_hash_info(algo);
+		if (!md_info) {
+			res =  TEE_ERROR_NOT_SUPPORTED;
+			goto err;
+		}
+
+		mbedtls_md_init(ctx);
+		lmd_res = mbedtls_md_setup(ctx, md_info, 1);
+		if (lmd_res != 0) {
+			FMSG("md setup failed, res is 0x%x", -lmd_res);
+			res = get_tee_result(lmd_res);
+			goto err;
+		}
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+#endif
+	default:
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto err;
+	}
+
+	*ctx_ret = ctx;
+	return TEE_SUCCESS;
+err:
+	free(ctx);
+	return res;
 }
 
-void crypto_mac_free_ctx(void *ctx, uint32_t algo __unused)
+void crypto_mac_free_ctx(void *ctx, uint32_t algo __maybe_unused)
 {
+	size_t ctx_size __maybe_unused;
+
+	/*
+	 * Check that it's a supported algo, or crypto_mac_alloc_ctx()
+	 * could never have succeded above.
+	 */
 	if (ctx)
-		assert(0);
+		assert(!mac_get_ctx_size(algo, &ctx_size));
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		mbedtls_md_free(ctx);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC_MAC)
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		break;
+#endif
+#if defined(CFG_CRYPTO_CMAC)
+	case TEE_ALG_AES_CMAC:
+		break;
+#endif
+	default:
+		break;
+	}
+	free(ctx);
 }
 
-void crypto_mac_copy_state(void *dst_ctx __unused, void *src_ctx __unused,
-			   uint32_t algo __unused)
+void crypto_mac_copy_state(void *dst_ctx, void *src_ctx, uint32_t algo)
 {
+#if defined(CFG_CRYPTO_HMAC) || defined(CFG_CRYPTO_CMAC)
+	int lmd_res;
+#endif
+
+	switch (algo) {
+#if defined(CFG_CRYPTO_HMAC)
+	case TEE_ALG_HMAC_MD5:
+	case TEE_ALG_HMAC_SHA224:
+	case TEE_ALG_HMAC_SHA1:
+	case TEE_ALG_HMAC_SHA256:
+	case TEE_ALG_HMAC_SHA384:
+	case TEE_ALG_HMAC_SHA512:
+		lmd_res = mbedtls_md_clone(dst_ctx, src_ctx);
```

Comment:
If this function fails we need to do something about it. I'd prefer changing `crypto_mac_copy_state()` to allow returning an error over silently ignoring the problem or calling `panic()`.

Reply:
Yes, I agree with you, I pref to call `panic()` here for least changing.